scene/gui/text_edit.cpp

```diff
@@ -4504,6 +4515,94 @@ int TextEdit::get_caret_count() const {
 	return carets.size();
 }
 
+void TextEdit::insert_caret_at_carets(bool p_below) {
+	// Stores all information of the new carets
+	// x = line, y = column, z = selection from column, w = selection to column
+	Vector<Vector4i> new_carets;
+	const int text_tab_size = text.get_tab_size();
+
+	new_carets.resize(caret_index_edit_order.size());
+	for (const int &caret_index : caret_index_edit_order) {
+		const int caret_line = get_caret_line(caret_index);
+
+		int caret_from_column, caret_to_column;
+		bool caret_is_at_to_column = true;
+		if (has_selection(caret_index)) {
+			// If the selection goes over multiple lines, deselect it.
+			if (get_selection_from_line(caret_index) != get_selection_to_line(caret_index)) {
+				deselect(caret_index);
+				caret_from_column = caret_to_column = get_caret_column(caret_index);
+			} else {
+				caret_from_column = get_selection_from_column(caret_index);
+				caret_to_column = get_selection_to_column(caret_index);
+				caret_is_at_to_column = get_caret_column(caret_index) == caret_to_column;
+			}
+		} else {
+			caret_from_column = caret_to_column = get_caret_column(caret_index);
+		}
+
+		// Get the columns of the caret with tab sizes in mind
+		const String text_line = get_line(caret_line);
+		const int pre_caret_from_column = caret_from_column;
+		const int pre_caret_to_column = caret_to_column;
+		for (int i = 0; i < pre_caret_to_column && i < text_line.length(); i++) {
+			if (text_line.get(i) == '\t') {
+				if (i < pre_caret_from_column) {
+					caret_from_column += text_tab_size - 1;
+				}
+				caret_to_column += text_tab_size - 1;
+			}
+		}
+
+		const int new_caret_line = p_below ? (caret_line + 1) : (caret_line - 1);
+		const String new_caret_text_line = get_line(new_caret_line);
+
+		// Count the number of columns of the line with tab sizes in mind
+		int new_caret_from_column = 0, new_caret_to_column = 0;
+		int from_tab_count = 0, to_tab_count = 0;
+		for (int j = 0; new_caret_to_column < caret_to_column && j < new_caret_text_line.length(); j++) {
+			const int char_size = new_caret_text_line.get(j) == '\t' ? text_tab_size : 1;
+			new_caret_to_column += char_size;
+			to_tab_count += char_size == text_tab_size;
+			if (new_caret_from_column < caret_from_column) {
+				new_caret_from_column += char_size;
+				from_tab_count += char_size == text_tab_size;
+			}
+		}
```

Comment:
This will not work with mixed direction / RTL text, you should not expect displayed text to be in the same order or have the same number of characters as the source string.


https://user-images.githubusercontent.com/7645683/195176782-26e234a6-83f2-4356-916e-20fb2c9d6249.mov

See how `_move_caret_down`/`up` work, it's doing it correctly.

Reply:
[control.gd.zip](https://github.com/godotengine/godot/files/9759262/control.gd.zip)

Reply:
I took the code from `set_caret_line`, put it into `_get_new_caret_line_column` and stripped it of everything I didn't need. Therefore this now supports everything `set_caret_line` and therefore `_move_caret_up`/`down` supports.